rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+package as dependency. This means any change to TF will take at most 24
+hours to be reflected on the Keras side.
+* The Keras code will be mirrored to a Google-internal code repository via Google-internal 
+tools within a very short time window after each change.
+The Google-internal CI tests will run on HEAD for both Keras and TF code.
+* The CI build for the repository on GitHub might break when it sees a
+new version of `tf-nightly`, if  certain behavior has been changed and wasn't
+caught by unit tests. We have  observed a few similar cases with
+[tf/addons](https://github.com/tensorflow/addons).
+We hope this can be reduced by stronger
+unit test coverage by Google internel systems, when both TF and Keras code are 
+tested at HEAD.
+* pip package management. Keras will now follow the `tf-estimator` approach. 
+"pip install tensorflow" should also install Keras (from PyPI) as well.
+There are more details for the pip package in the
+[Improved pip package structure](https://github.com/tensorflow/community/pull/182) RFC.
+
+### Dependency Cleanup
+
+As the high-level API of TensorFlow, Keras should have a direct dependency on
+TF low-level APIs, but not the other way around. Unfortunately, there is some existing reverse 
+logic in the TF code that relies on Keras, which we should update/remove 
+when we split the repository.
+
+The current usage of Keras from TensorFlow are:
+* Unit tests, which should be converted to integration tests, or port the tests
+to Keras repository.
+* `feature_column`.
+* Legacy `tf.layers` in v1 API.
+* legacy RNN cells.
+* TPU support code for `optimizer_v2`.
+* SavedModel.
+* TF Lite.
+
+All Keras imports in integration tests can be changed to use dynamic import like below:
+
+```python
+try:
+   from tensorflow.python.keras.engine import base_layer
+except ImportError:
+   tf.logging.error('keras is not installed, please pip install keras')
+   base_layer = None
+```
+
+### Update Keras to only use public TF APIs
+
+The current Keras code will still work if we do e.g.:
+```python
+from tensorflow.python.ops import array_ops
+
+ones = array_ops.ones([2, 3])
+```
+
+However, since Keras is a separate repository, having it only use TF 
+public APIs will heavily reduce the chance of breakage caused by relying
+on private methods or implementation details. We think this point is
+critial to the health of the project. This also allows TF to change internal 
+implementation details without worrying about breaking Keras.
+
+The converted code should look like e.g.:
+
+```python
+import tensorflow as tf
+
+ones = tf.ones([2, 3])
+```
+
+During this conversion, we might notice that certain TF features used in Keras are
+not public. A decision should be made on a case-by-case basis:
+
+* Copy the functionality from TF to Keras.
+* Replace the usage with another alternative TF public API.
+* Make the functionality a new TF public API.
+
+**Note that the open-source community is encouraged to contribute to this effort.**
```

Comment:
Very nice description of the problem. Since there is a lot of decision making going on there, I believe it would be nice for people in charge of decisions to open issues describing what they want to do for all the unwanted imports we'll have. It's hard to help when you don't know what will be decided to resolve the problem (here option 1? 2? or 3?). 

If we don't do that, I believe the open-source community won't help much just because of a lack of guidance.